R/plot.data_table_threads_benchmark.R

```diff
@@ -23,39 +23,49 @@
-  maxSpeedup <- df[, .(threadCount = threadCount[which.max(speedup)], speedup = max(speedup)), by = expr]
-  idealSpeedup <- seq(1, getDTthreads())
-  idealSpeedupData <- data.frame(threadCount = 1:getDTthreads(), speedup = idealSpeedup)
-  subOptimalSpeedupData <- data.frame(threadCount = seq(1, getDTthreads(), length.out = getDTthreads()), speedup = seq(1, getDTthreads()/2, length.out = getDTthreads()))
+  maxSpeedup <- dt[, .(threadCount = threadCount[which.max(speedup)], speedup = max(speedup)), by = expr]
 
-  closestPoints <- data.frame()
-  for(i in unique(df$expr)) {
-    dfSubset <- df[df$expr == i, ]
-    suboptimalSubset <- subOptimalSpeedupData[subOptimalSpeedupData$threadCount %in% dfSubset$threadCount, ]
-    closestPoint <- dfSubset[which.max(dfSubset$speedup - suboptimalSubset$speedup), ]
-    closestPoints <- rbind(closestPoints, closestPoint)
-  }
+  idealSpeedup.list <- lapply(unique(dt$expr), function(expr)
+  {
+    data.table(threadCount = 1:getDTthreads(),
+               speedup = seq(1, getDTthreads()),
+               expr = expr, type = "Ideal")
+  })
+  subOptimalSpeedup.list <- lapply(unique(dt$expr), function(expr)
+  {
+    data.table(threadCount = seq(1, getDTthreads(), length.out = getDTthreads()),
+               speedup = seq(1, getDTthreads()/2, length.out = getDTthreads()),
+               expr = expr, type = "Sub-optimal")
+  })
 
-  ggplot(df, aes(x = threadCount, y = speedup)) +
-    geom_line(aes(linetype = "Measured")) +
-    geom_line(data = idealSpeedupData, aes(x = threadCount, y = speedup, linetype = "Ideal"), color = "black") +
-    geom_line(data = subOptimalSpeedupData, aes(x = threadCount, y = speedup, linetype = "Sub-optimal"), color = "black") +
-    geom_point(data = closestPoints, aes(x = threadCount, y = speedup, shape = "Recommended"), color = "black", size = 2) +
-    geom_point(data = maxSpeedup, aes(x = threadCount, y = speedup, shape = "Best performing"), color = "red", size = 2) +
-    geom_text(data = closestPoints, aes(label = threadCount), vjust = -0.5, size = 4, na.rm = TRUE) +
-    geom_text(data = maxSpeedup, aes(label = threadCount), vjust = -0.5, size = 4, na.rm = TRUE) +
+  combinedLineData <- rbindlist(c(idealSpeedup.list, subOptimalSpeedup.list, list(dt)), use.names = TRUE, fill = TRUE)
+
+  closestPoints <- dt[, {
+    suboptimalSpeedupSubset <- subOptimalSpeedup.list[[.GRP]][threadCount %in% .SD$threadCount, on = .(threadCount)]
+    .SD[which.max(speedup - suboptimalSpeedupSubset$speedup)]
+  }, by = expr]
+
+  closestPoints[, medianTime := NULL]
+  closestPoints[, type := "Recommended"]
+  maxSpeedup[, type := "Best performing"]
+  combinedPointData <- rbind(maxSpeedup, closestPoints)
+
+  ggplot(dt, aes(x = threadCount, y = speedup)) +
+    geom_line(data = combinedLineData, aes(x = threadCount, y = speedup, linetype = type), color = "black") +
+    geom_point(data = combinedPointData, aes(x = threadCount, y = speedup, shape = type, color = type), size = 2) +
```

Comment:
please only use color and not linetype/shape, as we discussed last week

Reply:
Based on our conversation, I remember you mentioned to not use both color and shape for the lines as that is redundant, and instead only either one of them but such that we would still keep different legends for point and line geometry. So what you meant then is that I should just use different colors for *both* points and lines? (which would also means no 'Speedup'/'Thread Count' legends for lines and points respectively, noh? Also, I remember you mentioning to improve the wording for those two separately so I'm confused here) 

Something like this perhaps: (?)
<img width="810" alt="image" src="https://github.com/Anirban166/data.table.threads/assets/30123691/8f8c9b69-bbda-4249-96fb-054c2022af9a">

(I went ahead with what you suggested here by making some changes locally, and that's a plot from a test run with my updated code)

I got done with the other points in this review and pushed those updates - Thanks for the helpful suggestions!

Reply:
Based on our conversation, I remember you mentioned to not use both color and shape for the lines as that is redundant, and instead only either one of them but such that we would still keep different legends for point and line geometry. So what you meant then is that I should just use different colors for *both* points and lines? (which would also mean no 'Speedup' and 'Thread Count' in the legend for lines and points respectively, right? Also, I remember you mentioning to improve the wording for those two separately so I'm confused here) 

Something like this perhaps: (?)
<img width="810" alt="image" src="https://github.com/Anirban166/data.table.threads/assets/30123691/8f8c9b69-bbda-4249-96fb-054c2022af9a">

(I went ahead with what you suggested here by making some changes locally, and that's a plot from a test run with my updated code)

I got done with the other points in this review and pushed those updates - Thanks for the helpful suggestions!

Reply:
please change the data values to make it clear which line matches which point
- change Sub-optimal to Recommended, both should be the same color (blue?)
- change Best performing to Ideal, both should be  the same color (red?)